### PR TITLE
FDP-2551: Implement sending firmware update to device

### PR DESCRIPTION
**Pull request by sanderv**

### CODE REVIEW

File: components/psk/src/main/kotlin/org/gxf/crestdeviceservice/psk/service/PskService.kt [hunk 1]
```diff
@@ -26,17 +26,17 @@ class PskService(private val pskRepository: PskRepository, private val pskConfig
...
 
-    private fun getCurrentActivePsk(deviceId: String) =
+    fun getCurrentActivePsk(deviceId: String) =
         pskRepository.findFirstByIdentityAndStatusOrderByRevisionDesc(deviceId, PreSharedKeyStatus.ACTIVE)
+            ?: throw NoExistingPskException("No active PSK found")
 
     fun getCurrentPendingPsk(deviceId: String) =
         pskRepository.findFirstByIdentityAndStatusOrderByRevisionDesc(deviceId, PreSharedKeyStatus.PENDING)
 
     fun getCurrentReadyPsk(deviceId: String): PreSharedKey? {
```
Comment (loesimmens) [inline components/psk/src/main/kotlin/org/gxf/crestdeviceservice/psk/service/PskService.kt:38]: hier ook de NoExistingPskException gooien?

File: application/src/test/kotlin/org/gxf/crestdeviceservice/controller/PskControllerTest.kt [hunk 2]
```diff
@@ -3,10 +3,11 @@
 // SPDX-License-Identifier: Apache-2.0
 package org.gxf.crestdeviceservice.controller
 
+import org.gxf.crestdeviceservice.psk.exception.NoExistingPskException
 import org.gxf.crestdeviceservice.psk.service.PskService
 import org.gxf.crestdeviceservice.service.MetricService
 import org.junit.jupiter.api.Test
-import org.mockito.kotlin.whenever
+import org.mockito.kotlin.*
```
Comment (loesimmens) [inline application/src/test/kotlin/org/gxf/crestdeviceservice/controller/PskControllerTest.kt:10]: opslitsen in losse imports

File: application/src/test/kotlin/org/gxf/crestdeviceservice/service/DownlinkServiceTest.kt [hunk 3]
```diff
@@ -3,56 +3,82 @@
...
+import org.gxf.crestdeviceservice.model.Downlink
 import org.gxf.crestdeviceservice.psk.service.PskService
+import org.gxf.crestdeviceservice.service.command.CommandGenerator
+import org.junit.jupiter.api.BeforeEach
 import org.junit.jupiter.api.Test
-import org.junit.jupiter.params.ParameterizedTest
-import org.junit.jupiter.params.provider.CsvSource
-import org.mockito.kotlin.mock
-import org.mockito.kotlin.whenever
+import org.mockito.kotlin.*
```
Comment (loesimmens) [inline application/src/test/kotlin/org/gxf/crestdeviceservice/service/DownlinkServiceTest.kt:17]: imports opsplitsen

File: components/base85/README.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# Base85 Encoder/Decoder #
+
+High speed, standalone, thread-safe [Base85](https://en.wikipedia.org/wiki/Ascii85) encoder/decoder implemented in Java and JavaScript.
+
+Supports [RFC 1924](https://tools.ietf.org/html/rfc1924), [Z85](https://rfc.zeromq.org/spec:32/Z85/), and [Ascii85](https://en.wikipedia.org/wiki/Ascii85) (with optional null and space compression).
```
Comment (loesimmens) [inline components/base85/README.md:5]: Alleen RFC 1924, toch?

File: components/firmware/src/test/kotlin/org/gxf/crestdeviceservice/firmware/entity/FirmwareFactory.kt [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// SPDX-FileCopyrightText: Copyright Contributors to the GXF project
+//
+// SPDX-License-Identifier: Apache-2.0
+package org.gxf.crestdeviceservice.firmware.entity
+
+import java.util.*
```
Comment (loesimmens) [inline components/firmware/src/test/kotlin/org/gxf/crestdeviceservice/firmware/entity/FirmwareFactory.kt:6]: imports splitsen

File: components/firmware/src/test/kotlin/org/gxf/crestdeviceservice/firmware/entity/FirmwarePacketTest.kt [hunk 6]
```diff
@@ -0,0 +1,27 @@
+// SPDX-FileCopyrightText: Copyright Contributors to the GXF project
+//
+// SPDX-License-Identifier: Apache-2.0
+package org.gxf.crestdeviceservice.firmware.entity
+
+import java.util.*
```
Comment (loesimmens) [inline components/firmware/src/test/kotlin/org/gxf/crestdeviceservice/firmware/entity/FirmwarePacketTest.kt:6]: imports splitsen

File: components/firmware/src/test/resources/firmware-oneliner.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+OTA0000++%oAafJ@bH$e-l!rv8lPdLW9=tvX*t_nHb!L6~Z5Iz6^B2Y02EHN<!`0$7$5%6H+oDgV001uFe;E>=9l&2lflHZc=k{S@ussI200000J0RjUA0RRC1{{UJL(YgQt000000000203ZPu009Uh5Yfs207*_pPdZdZOaK4?000003J~6^04y*pFf1`KEFv*7Ga?NDLC(Cj<ktx$Tg*MYs57Jqc6pCjM_Km)VhHdIq3TwMa9I:DONE
+
+
+Bastiaan & Jeffrey
```
Comment (loesimmens) [inline components/firmware/src/test/resources/firmware-oneliner.txt:4]: ?

File: components/firmware/src/main/kotlin/org/gxf/crestdeviceservice/firmware/service/FirmwareHashService.kt [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+     *
+     * @param firmwarePacket the original packet (as an entity)
+     * @param deviceSecret the secret with which to hash the firmware hash(es)
+     * @return the downlink command ready to be sent to the device
+     */
+    fun generateDeviceSpecificPacket(firmwarePacket: FirmwarePacket, deviceSecret: String): String
+}
+
+@Service
+internal class InternalFirmwareHashService : FirmwareHashService {
```
Comment (loesimmens) [inline components/firmware/src/main/kotlin/org/gxf/crestdeviceservice/firmware/service/FirmwareHashService.kt:28]: Internal?

File: components/firmware/src/main/kotlin/org/gxf/crestdeviceservice/firmware/service/FirmwareHashService.kt [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            result = replaceCurrentFirmwareHash(result, deviceSecret)
+        }
+        if (firmwarePacket.isLastPacket()) {
+            result = replaceTargetFirmwareHash(result, deviceSecret)
+        }
+        return result
+    }
+
+    private fun replaceCurrentFirmwareHash(packet: String, deviceSecret: String): String {
+        val command = packet.substring(0, OTA_START.length)
```
Comment (loesimmens) [inline components/firmware/src/main/kotlin/org/gxf/crestdeviceservice/firmware/service/FirmwareHashService.kt:51]:
```suggestion
        val command = OTA_START
```

**Review by loesimmens — CHANGES_REQUESTED** [review]

File: components/firmware/src/main/kotlin/org/gxf/crestdeviceservice/firmware/service/FirmwareHashService.kt (hunk 8 above)
Comment (sanderv) [reply components/firmware/src/main/kotlin/org/gxf/crestdeviceservice/firmware/service/FirmwareHashService.kt:28]: Seemed like a good idea at the time... Removed

File: application/src/main/kotlin/org/gxf/crestdeviceservice/consumer/IncomingDeviceCredentialsConsumer.kt [hunk 10]
```diff
@@ -72,6 +72,6 @@ class IncomingDeviceCredentialsConsumer(
                 commandValue = null,
             )
 
-        commandService.saveCommandEntities(listOf(pskCommand, pskSetCommand))
+        commandService.save(listOf(pskCommand, pskSetCommand))
```
[the remark this answers is not in the document]
Comment (sanderv) [reply application/src/main/kotlin/org/gxf/crestdeviceservice/consumer/IncomingDeviceCredentialsConsumer.kt:75]: Combining also changes the return type for single params to a list of objects, unwanted side effect. I do like the vararg when calling a function, so changed it from List<> to vararg

File: application/src/main/kotlin/org/gxf/crestdeviceservice/service/DownlinkService.kt [hunk 11]
```diff
@@ -48,19 +50,15 @@ class DownlinkService(
             else -> true
         }
 
-    private fun getDownlinkFromCommands(deviceId: String, pendingCommands: List<Command>): String {
-        logger.info { "Device $deviceId has pending commands of types: ${getPrintableCommandTypes(pendingCommands)}." }
-
-        val downlink = Downlink(messageProperties.maxBytes)
+    private fun fillDownlinkFromCommands(pendingCommands: List<Command>, downlink: Downlink) {
```
[the remark this answers is not in the document]
Comment (sanderv) [reply application/src/main/kotlin/org/gxf/crestdeviceservice/service/DownlinkService.kt:53]: The downlink is filled by calling `addIfPossible(...)`

File: application/src/main/kotlin/org/gxf/crestdeviceservice/service/PayloadService.kt [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    ) {
+        val fotaMessageCounter = deviceMessage.getFotaMessageCounter()
+        if (fotaMessageCounter > 0) {
+            val firmwareCommand =
+                commandService.getAllCommandsInProgressForDevice(identity).find {
+                    it.type == Command.CommandType.FIRMWARE
+                }
+
+            if (firmwareCommand != null) {
+                val firmware = firmwareService.findByName(firmwareCommand.commandValue!!)
```
[the remark this answers is not in the document]
Comment (sanderv) [reply application/src/main/kotlin/org/gxf/crestdeviceservice/service/PayloadService.kt:49]: The error is already in the `Command.init {}`, so this "should never happen"

File: application/src/main/kotlin/org/gxf/crestdeviceservice/service/command/CommandGenerator.kt [hunk 13]
```diff
@@ -0,0 +1,15 @@
+// SPDX-FileCopyrightText: Copyright Contributors to the GXF project
+//
+// SPDX-License-Identifier: Apache-2.0
+package org.gxf.crestdeviceservice.service.command
```
[the remark this answers is not in the document]
Comment (sanderv) [reply application/src/main/kotlin/org/gxf/crestdeviceservice/service/command/CommandGenerator.kt:4]: Moved the generators to that package

File: components/base85/README.md (hunk 4 above)
Comment (sanderv) [reply components/base85/README.md:5]: This is the original README, included as-is. Not sure what to do here

**Review by levi-h — APPROVED** [review]

File: components/firmware/src/main/kotlin/org/gxf/crestdeviceservice/firmware/service/FirmwareService.kt [hunk 14]
```diff
@@ -3,7 +3,33 @@
...
 package org.gxf.crestdeviceservice.firmware.service
 
+import org.gxf.crestdeviceservice.device.service.DeviceService
+import org.gxf.crestdeviceservice.firmware.entity.Firmware
+import org.gxf.crestdeviceservice.firmware.repository.FirmwarePacketRepository
 import org.gxf.crestdeviceservice.firmware.repository.FirmwareRepository
 import org.springframework.stereotype.Service
 
-@Service class FirmwareService(private val firmwareRepository: FirmwareRepository)
+@Service
```
Comment (loesimmens) [inline components/firmware/src/main/kotlin/org/gxf/crestdeviceservice/firmware/service/FirmwareService.kt:12]: Comparison seems to be with an older version of main. Will merging this pr overwrite te existing function `processFirmware` in main? See https://github.com/OSGP/sng-crest-device-service/blob/main/components/firmware/src/main/kotlin/org/gxf/crestdeviceservice/firmware/service/FirmwareService.kt

**Review by loesimmens — CHANGES_REQUESTED** [review]

**Review by loesimmens — APPROVED** [review]

**Pull request merged**